nx/test/nx/defn/evaluator_test.exs

```diff
@@ -620,4 +620,52 @@ defmodule Nx.Defn.EvaluatorTest do
       assert cond_nested_condition_cache(%{iteration: 12}) == %{iteration: Nx.tensor(13)}
     end
   end
+
+  describe "vectorization" do
+    defn vectorized_while(t) do
+      while({t, i = 0}, i < 2) do
+        {t + 1, i + 1}
+      end
+    end
+
+    defn vectorized_cond(cond_1, on_1, cond_2, on_2, on_3) do
+      cond do
+        cond_1 -> on_1
+        cond_2 -> on_2
+        true -> on_3
+      end
+    end
+
+    defn vectorized_metadata(t) do
+      x = Nx.devectorize(t)
+      x = Nx.vectorize(x, t.vectorized_axes)
+
+      stop_grad(x)
+    end
+
+    test "while" do
+      t = Nx.iota({2, 3}, vectorized_axes: [a: 1])
+
+      assert {result, i} = vectorized_while(t)
+
+      assert result == Nx.add(t, 2)
+      assert i == Nx.tensor(2)
+    end
+
+    test "cond" do
+      a = Nx.tensor(3)
+      b = Nx.tensor(2)
+      c = Nx.vectorize(Nx.tensor([[1]]), x: 1, y: 1)
+
+      # This shows that non-vectorized branches still return unvectorized.
+      assert Nx.add(c, 2) == vectorized_cond(1, a, 0, b, c)
+      assert Nx.add(c, 1) == vectorized_cond(0, a, 1, b, c)
+      assert c == vectorized_cond(0, a, 0, b, c)
+    end
```

Comment:
This test is failing now. Any ideas @josevalim ?

Reply:
apart from this and the container thing above, everything seems to be working correctly

Reply:
I think it should fail. All branches should return the same shape because the branches are executed at compile time and, at compile time, you don't know which branch will win. We should probably raise if they are not equally vectorized.

Reply:
For while we have a similar issue. What happens if `initial` is vectorized and its return from the while `body` is not? It means the value is changing and becoming incompatible.

So I think, ultimately, we want to make sure they are compatible, and only devectorize if they are all equally vectorized!